rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+
+## Alternatives
+
+- Keep the status quo of not having an official formatter.
+  The danger is that this creates discord within the Nix community.
+  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
+- Specify a different format.
+- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
+- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
+  However, such an approach would not reduce the amount of merge conflicts,
+  and increase the workload for contributors during that time significantly.
+
+## Unresolved questions
+
+- Should the line length limit be specified?
+
+## Future work
+
+- General style guidelines beyond AST reformatting
+- Making widespread use of linters like Nixpkgs-hammering
+- Enforcing the format to other official repositories
+
+-----
+
+## Initial standard Nix format
+
+Terms and definitions:
+- Brackets: `[]`
+- Braces: `{}`
+- Parentheses: `()`
+
+- Newlines are not guaranteed to be preserved, but empty lines are.
+  - This allows the formatter to compact down multi-line expressions if necessary, while still allowing to structure the code appropriately.
+- Expressions of the same kind that can be treated as a sequence of expressions on the same level should be treated as such, even though they are technically parsed as a nested tree.
+  - This applies to else-if chains, functions with multiple arguments, some operators, etc.
+  - Example:
+    ```nix
+    # This is treated as a sequence of if-then-elsa's chains
+    if cond1 then
+      foo
+    else if cond2 then
+      bar
+    else
+      baz
+    ```
+
+### Single-line common ancestor expression rule
+
+For any two (sub-)expressions that are fully on a common single line, their smallest common ancestor expression must also be on the same line.
+
+**Example**
+
+```nix
+# Bad, expressions cond and bar are fully on the same line,
+# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
+if cond then foo
+else bar
+
+# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
+# which is also fully on the same line
+if cond then foo else bar
+
+# Bad, due to function application precedence, the smallest common ancestor expression
+# of foo and bar is `foo || bar baz`, which spans two lines
+foo || bar
+  baz
+```
+
+**Rationale**
+
+This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.
+
+### Line length
+
+Lines have a soft length limit that doesn't count indentation.
+There may also be a hard length limit that includes indentation.
+String-like values such as strings, paths, comments, urls, etc. may go over the hard length limit.
+These limits should be configurable, and the soft length limit should default to 100 characters.
```

Comment:
```suggestion
These limits should be configurable, and the soft length limit should default to 80 characters.
```

I am curious to know how much this was already discussed. Finding the right compromise for line length is essential.
It ensures the readability of expressions (also on smaller screen sizes) but doesn't affect the filesize significantly. 
While it might be less controversial to choose a larger line length limit (although it is a soft one), 

**I propose that we choose a smaller limit (such as 80 characters) and follow other established conventions, such as those in rust, python, or javascript, who really dug into that problem already.**

If it is configurable anyway it should be more strict and people who don't like that can change it


Rustfmt states, like most other popular formatting tools.

> Lines should be at most 100 characters. **It's even better if you can keep things to 80.**

Also prettier states

> For readability we recommend against using more than 80 characters:

- nix only uses two '  ' spaces for indentation, which is relatively low. 
  > Why should we have a higher line limit if indentation takes less space?
  > Keeping lines short and concise is a good thing as other tools have already been discovered.
  
  
- [rust](https://rustc-dev-guide.rust-lang.org/conventions.html#:~:text=Lines%20should%20be%20at%20most%20100%20characters.)
- [python](https://peps.python.org/pep-0008/#:~:text=The%20Python%20standard%20library%20is,docstrings%2Fcomments%20to%2072).
- [javascript](https://prettier.io/docs/en/options.html)

Reply:
You can have a look for yourself: https://github.com/piegamesde/nixpkgs/pull/3

Opinions on line lengths vary wildly, from people who want to go back to 80, to people who say the code looks best at ~130. But so far I've seen mostly positive feedback on the change from 80 to 100.

Also note that Nix ist not Rust or Python or any other language. Due to its nature, code often times contains long strings. We need to account for that in our decisions.

Reply:
I think a good rule of thumb would be half the horizontal width of a reasonably large monospaced font size (14-15) on a 16:9 monitor.

Reply:
@Atemu Interesting heuristic, just counted for mine and got 98 characters for my screen and a comfortable font size! Though if you add vim sidebars and file trees and co. that shrinks..

Reply:
Here's the diff on Nixpkgs when changing the default line limit from 100 to 80: https://github.com/piegamesde/nixpkgs/commit/e345d638168c066fae7e5bcfee414eae919712b1

Reply:
> Here's the diff on Nixpkgs when changing the default line limit from 100 to 80: [piegamesde/nixpkgs@e345d63](https://github.com/piegamesde/nixpkgs/commit/e345d638168c066fae7e5bcfee414eae919712b1)

On my phone the shorter line length improves readability because fewer lines are wrapped.

Reply:
> Here's the diff on Nixpkgs when changing the default line limit from 100 to 80: https://github.com/piegamesde/nixpkgs/commit/e345d638168c066fae7e5bcfee414eae919712b1

I definitely prefer the 80s limit. And would recommend it to the rfc decision. I know nix is not any other language but smaller limits exist for many reasons. Having large limits just shows that someone cannot format his code properly. It makes code less readable if you squish more expressions and statements into a single line.
- Having exceptions though for long paths or single line strings is fine i think. 

Controversial topic

Reply:
Note that this is only the _default_ line width, it's specified to be configurable.

Personally I don't have a strong opinion on this, 80 sounds fine too.

Reply:
I'm opposed to an 80 character limit, but in favor of an 80 character recommendation with a 100 character default limit.

However far beyond that, @hsjobeki I am strongly opposed to insulting people's skill on a code style generalization, and I also would appreciate it if you didn't hinder inclusivity with male-default language. Thank you.

Reply:
Please remember that this is about the format to be used by Nixpkgs. That's the value which matters to me, regardless of whether it ends up matching the nixfmt CLI default or not.

Also please remember that this is about an *automatic formatter*, the goal is precisely that nobody should have to manually format any code anymore.

Reply:
I believe that a lot of newlines distract from the content of the code. Often a statement is one unit of logic and splitting it up makes it harder to follow what's really relevant.

Having fewer newlines packs the screen more densely with code so less scrolling is needed to get an overview and orientation.

80 character rules are very old from times with lower resolutions. Most people don't code on their phones.

I have also no very strong opinion on this but these are the arguments why I reallylike the change and advocate for keeping it.

Reply:
> - Having exceptions though for long paths or single line strings is fine i think.
> Controversial topic

[nope](https://github.com/piegamesde/nixpkgs/pull/4/commits/d6b557762ee83fb9cb4a56bf8f0fd3d01c791d2f)

----

Btw I'm currently experimenting with some more rules which allow to bypass the line length limit in some circumstances: https://github.com/piegamesde/nixpkgs/pull/4/files/eb3daa3..b247f67

There are some false positives, but overall I like the direction where this is going. (This will be discussed in the next meeting.)

Reply:
Discussed in the meeting today, we agree that the default of a 100-char soft limit is pretty good. We came to that conclusion already before when we went through various limits. But in any case, it's explicitly stated to be configurable, so if people have different preferences, they can adjust it as necessary. We think this doesn't need to be discussed further.

Reply:
Personally, I care about a hard limit much more than a soft limit. That way I can predictably avoid wrapping.

Reply:
@ursi You cannot avoid wrapping either way, because a lot of lines simply are longer than the line limit with no possibility of breaking them up.